resources/tentative/complex-static-html/src/big-dom-generator/gen-css.js

```diff
@@ -159,80 +121,108 @@ const randomWeighted = (options, probs) => {
     return options[options.length - 1];
 };
 
-const buildMatchingSelector = (depth, index, oldCombinator, selLen, maxLen) => {
+const buildMatchingSelector = (element, depth, oldCombinator, selLen, maxLen) => {
     // prettier-ignore
     if (selLen >= maxLen)
         return "";
 
-    const getSelector = randomWeighted([getClassname, getType, () => "*"], [0.6, 0.3, 0.1]);
-    const selector = getSelector(depth, index);
-    const combinator = chooseCombinator(depth, index);
-
+    // Get a random selector for the element.
+    const getSelector = randomWeighted([getClassname, getElementType, () => "*"], [0.6, 0.3, 0.1]);
+    const selector = getSelector(element);
     // prettier-ignore
     if (!depth)
         return `${selector}${oldCombinator}`;
 
+    const children = Array.from(element.parentElement.children);
+    const index = children.indexOf(element);
+    const combinator = chooseCombinator(depth, index);
+
     const nextDepth = getNextDepth(combinator, depth);
-    const nextIndex = getNextIndex(combinator, nextDepth, index);
-    return buildMatchingSelector(nextDepth, nextIndex, combinator, selLen + 1, maxLen) + selector + oldCombinator;
+    const nextElement = getElementAtDepth(combinator, element, depth, nextDepth);
+
+    // Recurse with the next element and depth, and append the selector and old combinator.
+    return buildMatchingSelector(nextElement, nextDepth, combinator, selLen + 1, maxLen) + selector + oldCombinator;
 };
 
-// Returns a non-matching selector for the element.
-// We kept the selector from matching by adding the `.just-span` class to the left-most selector or
-// by adding a classname from one of its children.
-// TODO: Is there a better way to ensure the selector is non-matching?
-const buildNonMatchingSelector = (depth, index, oldCombinator, selLen, badSelector) => {
-    // If we are in the top node, we are done.
+const buildNonMatchingSelector = (element, depth, oldCombinator, selLen, badSelector) => {
     // prettier-ignore
     if (!depth)
         return `.just-span${ oldCombinator}`;
 
-    // If we've reached the target length, pick a random classname from its children.
-    const getSelector = randomWeighted([getClassname, getType, () => "*"], [0.6, 0.3, 0.1]);
-    const selector = getSelector(depth, index);
+    const getSelector = randomWeighted([getClassname, getElementType, () => "*"], [0.6, 0.3, 0.1]);
+    const selector = getSelector(element);
     if (selLen === badSelector) {
-        const wrongDepth = random.randRange(Math.min(depth + 1, 7), 8);
-        const wrongSelector = getClassname(wrongDepth, 0);
-        return selector + wrongSelector + oldCombinator;
+        const wrongSelector = getClassname(random.choice(Array.from(element.children)));
+        return `${selector} ${wrongSelector}${oldCombinator}`;
```

Comment:
The space between `${selector}` and `${wrongSelector}` would make this a descendant selector, which is not supposed to be.